### ISSUE TITLE
Setting `class_weight` in `fit()` causes error: `AttributeError: 'DistributedDatasetsFromFunction' object has no attribute 'map'`

### ISSUE THREAD

**Issue by juzzmac** [intel-analytics/BigDL#5713]
Hi, I'm running into this error whenever I set the `class_weight` parameter. The demo script below replicates the error. If the `class_weight` parameter is removed then the code completes without issue:

```
from bigdl.orca.learn.tf2 import Estimator

from tensorflow.keras.layers import Dense, Input
from tensorflow.keras.models import Model

from bigdl.orca import init_orca_context, OrcaContext

init_orca_context(cluster_mode="local")
spark = OrcaContext.get_spark_session()


def model_creator(config):
    input_layer = Input(shape=(1,))
    x = Dense(2, activation='relu')(input_layer)
    output_layer = Dense(1, activation='relu')(x)

    model = Model(inputs=input_layer, outputs=output_layer)
    model.compile(
        loss='categorical_crossentropy',
        optimizer='adam',
        metrics=['accuracy']
    )
    return model


est = Estimator.from_keras(model_creator=model_creator, workers_per_node=2)

test_data = spark.createDataFrame(
    [
        {'VAR1': 5, 'VAR2': 0},
        {'VAR1': 6, 'VAR2': 0},
        {'VAR1': 7, 'VAR2': 0},
        {'VAR1': 9, 'VAR2': 1},
        {'VAR1': 10, 'VAR2': 1},
        {'VAR1': 11, 'VAR2': 1},
    ]
)

batch_size = 2

stats = est.fit(
    test_data,
    feature_cols=['VAR1'],
    label_cols=['VAR2'],
    epochs=1,
    batch_size=batch_size,
    steps_per_epoch=6 // batch_size,
    class_weight={0: 1, 1: 5},
)

```

**Comment by sgwhat** [comment]
Hi sorry for the later reply. We have reproduced and looked into this issue.

### Reason
This is a Tensorflow limit. Orca tf2 estimator will use `experimental_distribute_dataset ` strategy when processing spark dataframe input, and Tensorflow doesn't allow user to add any more transformations to a `tf.distribute.DistributedDataset`. You could refer to [tf strategy doc](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset) to see more details.

**Comment by jason-dai** [comment]
> Hi sorry for the later reply. We have reproduced and looked into this issue.
> 
> ### Reason
> This is a Tensorflow limit. Orca tf2 estimator will use `experimental_distribute_dataset ` strategy when processing spark dataframe input, and Tensorflow doesn't allow users to add any more transformations to a `tf.distribute.DistributedDataset`. So you met this error when specifying `class_weights` (which requires map function) in your program, you could remove it to complete the training pipeline.
> 
> You could refer to [tf strategy doc](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset) to see more details.

Why do we support `class_weight` argument in `TensorFlow2Estimator.fit` then?

Does Keras support `class_weight` when using Distributed Strategy?

**Comment by sgwhat** [comment]
> > Hi sorry for the later reply. We have reproduced and looked into this issue.
> > ### Reason
> > This is a Tensorflow limit. Orca tf2 estimator will use `experimental_distribute_dataset ` strategy when processing spark dataframe input, and Tensorflow doesn't allow users to add any more transformations to a `tf.distribute.DistributedDataset`. So you met this error when specifying `class_weights` (which requires map function) in your program, you could remove it to complete the training pipeline.
> > You could refer to [tf strategy doc](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset) to see more details.
> 
> Why do we support `class_weight` argument in `TensorFlow2Estimator.fit` then?
> 
> Does Keras support `class_weight` when using Distributed Strategy?

1. Maybe for the consistency with tensorflow fit.
2. Keras not support `class_weight` when using `experimental_distribute_dataset ` strategy.

**Comment by sgwhat** [comment]
> 

Yes, tensorflow has described this in the strategy doc I mentioned before.

**Comment by sgwhat** [comment]
> > > > Hi sorry for the later reply. We have reproduced and looked into this issue.
> > > > ### Reason
> > > > This is a Tensorflow limit. Orca tf2 estimator will use `experimental_distribute_dataset ` strategy when processing spark dataframe input, and Tensorflow doesn't allow users to add any more transformations to a `tf.distribute.DistributedDataset`. So you met this error when specifying `class_weights` (which requires map function) in your program, you could remove it to complete the training pipeline.
> > > > You could refer to [tf strategy doc](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset) to see more details.
> > > 
> > > 
> > > Why do we support `class_weight` argument in `TensorFlow2Estimator.fit` then?
> > > Does Keras support `class_weight` when using Distributed Strategy?
> > 
> > 
> > 
> > 1. Maybe for the consistency with tensorflow fit.
> > 2. Keras not support `class_weight` when using `experimental_distribute_dataset ` strategy.
> 
> Then this is really a limitation of distributed TensorFlow/Keras, yes?

Yes, tensorflow has described this in the strategy doc I mentioned before.

**Comment by jason-dai** [comment]
> > > > > Hi sorry for the later reply. We have reproduced and looked into this issue.
> > > > > ### Reason
> > > > > This is a Tensorflow limit. Orca tf2 estimator will use `experimental_distribute_dataset ` strategy when processing spark dataframe input, and Tensorflow doesn't allow users to add any more transformations to a `tf.distribute.DistributedDataset`. So you met this error when specifying `class_weights` (which requires map function) in your program, you could remove it to complete the training pipeline.
> > > > > You could refer to [tf strategy doc](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset) to see more details.
> > > > 
> > > > 
> > > > Why do we support `class_weight` argument in `TensorFlow2Estimator.fit` then?
> > > > Does Keras support `class_weight` when using Distributed Strategy?
> > > 
> > > 
> > > 
> > > 1. Maybe for the consistency with tensorflow fit.
> > > 2. Keras not support `class_weight` when using `experimental_distribute_dataset ` strategy.
> > 
> > 
> > Then this is really a limitation of distributed TensorFlow/Keras, yes?
> 
> Yes, tensorflow has described this in the strategy doc I mentioned before.

We should then remove this parameter

**Comment by sgwhat** [comment]
> 

Sure.

**Comment by sgwhat** [comment]
> > > > > > Hi sorry for the later reply. We have reproduced and looked into this issue.
> > > > > > ### Reason
> > > > > > This is a Tensorflow limit. Orca tf2 estimator will use `experimental_distribute_dataset ` strategy when processing spark dataframe input, and Tensorflow doesn't allow users to add any more transformations to a `tf.distribute.DistributedDataset`. So you met this error when specifying `class_weights` (which requires map function) in your program, you could remove it to complete the training pipeline.
> > > > > > You could refer to [tf strategy doc](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset) to see more details.
> > > > > 
> > > > > 
> > > > > Why do we support `class_weight` argument in `TensorFlow2Estimator.fit` then?
> > > > > Does Keras support `class_weight` when using Distributed Strategy?
> > > > 
> > > > 
> > > > 
> > > > 1. Maybe for the consistency with tensorflow fit.
> > > > 2. Keras not support `class_weight` when using `experimental_distribute_dataset ` strategy.
> > > 
> > > 
> > > Then this is really a limitation of distributed TensorFlow/Keras, yes?
> > 
> > 
> > Yes, tensorflow has described this in the strategy doc I mentioned before.
> 
> We should then remove this parameter

Sure.
